File: lib/lim/src/logical/Friendship.cpp
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2017 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

#include "lim/inc/lim.h"
#include "lim/inc/Common.h"
#include "common/inc/WriteMessage.h"

#include "lim/inc/messages.h"
#include <algorithm>
#include <string.h>
#include "common/inc/math/common.h"


namespace columbus { namespace lim { namespace asg {

typedef boost::crc_32_type  Crc_type;

namespace logical { 
  Friendship::Friendship(NodeId _id, Factory *_factory) :
    Base(_id, _factory),
    m_grants(0),
    m_friend(0)
  {
  }

  Friendship::~Friendship() {
  }

  void Friendship::clone(const Friendship& other, bool tryOnVirtualParent) {
    base::Base::clone(other, false);

    m_grants = other.m_grants;
    m_friend = other.m_friend;
  }

  void Friendship::prepareDelete(bool tryOnVirtualParent){
    removeFriend();
    base::Base::prepareDelete(false);
  }

  NodeKind Friendship::getNodeKind() const {
    return ndkFriendship;
  }

  Key Friendship::getGrantsKey() const {
    return m_grants;
  }

  const std::string& Friendship::getGrants() const {
    return factory->getStringTable().get(m_grants);
  }

  void Friendship::setGrantsKey(Key _grants) {
    m_grants = _grants;
  }

  void Friendship::setGrants(const std::string& _grants) {
    m_grants = factory->getStringTable().set(_grants);
  }

  logical::Scope* Friendship::getFriend() const {
    logical::Scope *_node = NULL;
    if (m_friend != 0)
      _node = dynamic_cast<logical::Scope*>(factory->getPointer(m_friend));
    if ( (_node == NULL) || factory->getIsFiltered(_node))
      return NULL;

    return _node;
  }

  bool Friendship::setEdge(EdgeKind edgeKind, NodeId edgeEnd, void* acValue, bool tryOnVirtualParent) {
    switch (edgeKind) {
      case edkFriendship_Friend:
        setFriend(edgeEnd);
        return true;
      default:
        break;
    }
    if (base::Base::setEdge(edgeKind, edgeEnd, acValue, false)) {
      return true;
    }
    return false;
  }

  bool Friendship::removeEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    switch (edgeKind) {
      case edkFriendship_Friend:
        removeFriend();
        return true;
      default:
        break;
    }
    if (base::Base::removeEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  void Friendship::setFriend(NodeId _id) {
    logical::Scope *_node = NULL;
    if (_id) {
      if (!factory->getExist(_id))
        throw LimException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

      _node = dynamic_cast<logical::Scope*> (factory->getPointer(_id));
      if ( _node == NULL) {
        throw LimException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);
      }
      if (&(_node->getFactory()) != this->factory)
        throw LimException(COLUMBUS_LOCATION, CMSG_EX_THE_FACTORY_OF_NODES_DOES_NOT_MATCH );

      if (Common::getIsBaseClassKind(_node->getNodeKind(), ndkMethod) || Common::getIsBaseClassKind(_node->getNodeKind(), ndkClass)) {
        if (m_friend) {
          if (factory->getExistsReverseEdges())
            factory->reverseEdges->removeEdge(m_friend, m_id, edkFriendship_Friend);
        }
        m_friend = _node->getId();
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->insertEdge(m_friend, this->getId(), edkFriendship_Friend);
      } else {
        throw LimException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);
      }
    } else {
      if (m_friend) {
        throw LimException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);
      }
    }
  }

  void Friendship::setFriend(logical::Scope *_node) {
    if (_node == NULL)
      throw LimException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);

    setFriend(_node->getId());
  }

  void Friendship::removeFriend() {
      if (m_friend) {
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_friend, m_id, edkFriendship_Friend);
      }
      m_friend = 0;
  }

  void Friendship::accept(Visitor &visitor) const {
    visitor.visit(*this);
  }

  void Friendship::acceptEnd(Visitor &visitor) const {
    visitor.visitEnd(*this);
  }

  double Friendship::getSimilarity(const base::Base& base){
    if(base.getNodeKind() == getNodeKind()) {
      const Friendship& node = dynamic_cast<const Friendship&>(base);
      double matchAttrs = 0;
      std::string str1, str2;
      size_t strMax;
      double strSim;
      str1 = getGrants();
      str2 = node.getGrants();
      strMax = std::max(str1.size(), str2.size());
      strSim = 1 - ((double)common::math::editDistance(str1, str2) / (strMax > 0 ? strMax : 1));
      if (strSim < Common::SimilarityMinForStrings)
        return 0.0;
      matchAttrs += strSim;
      return matchAttrs / (1 / (1 - Common::SimilarityMinimum)) + Common::SimilarityMinimum;
    } else {
      return 0.0;
    }
  }

  void Friendship::swapStringTable(RefDistributorStrTable& newStrTable, std::map<Key,Key>& oldAndNewStrKeyMap ){
    std::map<Key,Key>::iterator foundKeyId;

    foundKeyId = oldAndNewStrKeyMap.find(m_grants);
    if (foundKeyId != oldAndNewStrKeyMap.end()) {
      m_grants = foundKeyId->second;
    } else {
      Key oldkey = m_grants;
      m_grants = newStrTable.set(factory->getStringTable().get(m_grants));
      oldAndNewStrKeyMap.insert(std::pair<Key,Key>(oldkey,m_grants));    }

  }

  NodeHashType Friendship::getHash(std::set<NodeId>& travNodes) const {
    if (hashOk) return nodeHashCache;
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_BEGIN,this->getId()); 
    if (travNodes.count(getId())>0) {
      common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_SKIP);
      return 0;
    }
    travNodes.insert(getId());
    Crc_type resultHash;
    resultHash.process_bytes( "logical::Friendship", strlen("logical::Friendship"));
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_END,resultHash.checksum()); 
    nodeHashCache = resultHash.checksum();
    hashOk = true;
    return nodeHashCache;
  }

  void Friendship::save(io::BinaryIO &binIo,bool withVirtualBase  /*= true*/) const {
    Base::save(binIo,false);

    factory->getStringTable().setType(m_grants, StrTable::strToSave);
    binIo.writeUInt4(m_grants);

    binIo.writeUInt4(m_friend);

  }

  void Friendship::load(io::BinaryIO &binIo, bool withVirtualBase /*= true*/) {
    Base::load(binIo,false);

    m_grants = binIo.readUInt4();

    m_friend =  binIo.readUInt4();

  }


}


}}}

File: csharp/lib/lim/Nodes/Logical/Friendship.cs
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2017 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

using System;
using System.Text;
using System.Diagnostics;
using System.Collections.Generic;

namespace Columbus.Lim.Asg.Nodes.Logical {

    /// <summary>
    /// Friendship class, which represents the logical::Friendship node.
    /// </summary>
    [DebuggerDisplay("{DemangledName}")]
#if (LIM_INTERNAL)
    internal
#else
    public
#endif
    class Friendship : Columbus.Lim.Asg.Nodes.Base.Base {

        /// <summary>The Key of the `grants`.</summary>
        protected uint m_grants;


        // ---------- Edges ----------

        /// <summary>The id of the node the friend edge points to.</summary>
        protected uint m_friend;

        /// <summary>
        /// Constructor, only factory can instantiates nodes.
        /// </summary>
        /// <param name="nodeId">[in] The id of the node.</param>
        /// <param name="factory">[in] Poiter to the Factory the node belongs to.</param>
        public Friendship(uint nodeId, Factory factory) : base(nodeId, factory) {
            m_grants = 0;
            m_friend = 0;
        }

        /// <summary>
        /// Gives back the NodeKind of the node.
        /// </summary>
        /// <returns>Returns with the NodeKind.</returns>
        public override Types.NodeKind NodeKind {
            get { return Types.NodeKind.ndkFriendship; }
        }


        // ---------- Attribute propert(y)(ies) ----------

        /// <summary>
        /// Gives back or sets the grants of the node.
        /// </summary>
        public string Grants
        {
            get
            {
                return fact.StringTable.get(m_grants);
            }
            set
            {
                m_grants = fact.StringTable.set( value );
            }
        }

        /// <summary>
        /// Gives back or sets the Key of grants of the node.
        /// </summary>
        public uint GrantsKey
        {
            get
            {
                return m_grants;
            }
            set
            {
               m_grants = value;
            }
        }


        // ---------- Reflection getter function ----------

        /// <summary>
        /// Gives back the begin iterator for the edge specified
        /// </summary>
        /// <param name="edge">[in] The multiple edge for which the iterator is requested</param>
        public override ListIterator<Columbus.Lim.Asg.Nodes.Base.Base> GetListIteratorBegin(string edge) {
            switch(edge) {
                default:
                    return base.GetListIteratorBegin(edge);
            }
        }


        // ---------- Reflection add function ----------

        /// <summary>
        /// Adds item as a new edge
        /// </summary>
        public override void add(string edge, uint nodeId) {
            switch(edge) {
                default:
                    base.add(edge, nodeId);
                    return;
            }
        }


        // ---------- Edge getter function(s) ----------

        /// <summary>
        /// Gives back the id of the node the friend edge points to.
        /// </summary>
        /// <returns>Returns the end point of the m_friend edge.</returns>
        /// <exception cref="Columbus.Lim.Asg.LimException">Throws LimException if the edge id is invalid.</exception>
        public uint getFriend() {
            if (fact.getIsFiltered(m_friend))
                return 1;
            else
                return m_friend;
        }


        // ---------- Edge setter function(s) ----------

        /// <summary>
        /// Sets the friend edge.
        /// </summary>
        /// <param name="nodeId">[in] The new end point of the friend edge or 0. With 0 the edge can be deleted.</param>
        /// <exception cref="Columbus.Lim.Asg.LimException">Throws LimException if the nodeId is invalid.</exception>
        public void setFriend(uint nodeId) {
            if (nodeId != 0) {
                if (!fact.getExist(nodeId))
                    throw new LimException("Columbus.Lim.Asg.Nodes.Logical.Friendship.setFriend(NodeId)", "The end point of the edge does not exist.");

                Columbus.Lim.Asg.Nodes.Base.Base node = fact.getRef(nodeId);
                if (Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkMethod) || Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkClass)) {
                    if (m_friend != 0) {
                    }
                    m_friend = nodeId;
                } else {
                    throw new LimException("Columbus.Lim.Asg.Nodes.Logical.Friendship.setFriend(NodeId)", "Invalid NodeKind (" + node.NodeKind.ToString() + ").");
                }
            } else {
                if (m_friend != 0) {
                }
                m_friend = 0;
            }
        }

        /// <summary>
        /// Sets the friend edge.
        /// </summary>
        /// <param name="node">[in] The new end point of the friend edge.</param>
        /// <exception cref="Columbus.Lim.Asg.LimException">Throws LimException if there's something wrong with the given node.</exception>
        public void setFriend(Columbus.Lim.Asg.Nodes.Logical.Scope node) {
            if (Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkMethod) || Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkClass)) {
                if (m_friend != 0) {
                }
                m_friend = node.Id;
            } else {
                throw new LimException("Columbus.Lim.Asg.Nodes.Logical.Friendship.setFriend(NodeId)", "Invalid NodeKind (" + node.NodeKind.ToString() + ").");
            }
        }


        // ---------- Accept functions for Visitor ----------

        /// <summary>
        /// It calls the appropriate visit method of the given visitor.
        /// </summary>
        /// <param name="visitor">[in] The used visitor.</param>
        public override void accept(Visitors.Visitor visitor) {
            visitor.visit(this);
        }

        /// <summary>
        /// It calls the appropriate visitEnd method of the given visitor.
        /// </summary>
        /// <param name="visitor">[in] The used visitor.</param>
        public override void acceptEnd(Visitors.Visitor visitor) {
            visitor.visitEnd(this);
        }

        /// <summary>
        /// Saves the node.
        /// </summary>
        /// <param name="io">[in] The node is written into io.</param>
        /// <exception cref="Columbus.ColumbusIOException">Throws ColumbusIOException if there is something wrong with the file.</exception>
        public override void save(IO io) {
            base.save(io);

            fact.StringTable.setType(m_grants, StrTable.StrType.strToSave);
            io.writeUInt4(m_grants);

            io.writeUInt4(m_friend);

        }

        /// <summary>
        /// Loads the node.
        /// </summary>
        /// <param name="io">[in] The node is read from io.</param>
        /// <exception cref="Columbus.ColumbusIOException">Throws ColumbusIOException if there is something wrong with the file.</exception>
        public override void load(IO binIo) {
            base.load(binIo);

            m_grants = binIo.readUInt4();

            m_friend =  binIo.readUInt4();

        }

    }

}
